fido/Crypto/Fido2/PublicKey.hs

```diff
@@ -0,0 +1,279 @@
+    toAlg (-36) = pure $ ECDSAIdentifier ES512
+    toAlg (-8) = pure EdDSA
+    toAlg _ = fail "Unsupported `alg`"
+
+instance ToJSON COSEAlgorithmIdentifier where
+  toJSON (ECDSAIdentifier ES256) = Aeson.Number (-7)
+  toJSON (ECDSAIdentifier ES384) = Aeson.Number (-35)
+  toJSON (ECDSAIdentifier ES512) = Aeson.Number (-36)
+  toJSON EdDSA = Aeson.Number (-8)
+
+data EdDSAKey
+  = Ed25519 Ed25519.PublicKey
+  | Ed448 Ed448.PublicKey
+  deriving (Eq, Show)
+
+instance Arbitrary EdDSAKey where
+  arbitrary =
+    oneof
+      [ Ed25519 <$> randomKey1,
+        Ed448 <$> randomKey2
+      ]
+
+randomKey1 :: Gen Ed25519.PublicKey
+randomKey1 = do
+  rng <- drgNewSeed . seedFromInteger <$> arbitrary
+  let (a, _) = withDRG rng (Ed25519.toPublic <$> Ed25519.generateSecretKey)
+  pure a
+
+randomKey2 :: Gen Ed448.PublicKey
+randomKey2 = do
+  rng <- drgNewSeed . seedFromInteger <$> arbitrary
+  let (a, _) = withDRG rng (Ed448.toPublic <$> Ed448.generateSecretKey)
+  pure a
+
+randomKey3 :: Gen ECDSA.PublicKey
+randomKey3 = do
+  curve <- elements [ECC.SEC_p256r1, ECC.SEC_p384r1, ECC.SEC_p521r1]
+  rng <- drgNewSeed . seedFromInteger <$> arbitrary
+  let ((pub, _), _) = withDRG rng (ECC.generate (ECC.getCurveByName curve))
+  pure pub
+
+data ECDSAKey = ECDSAKey ECDSAIdentifier ECDSA.PublicKey deriving (Eq, Show)
+
+instance Arbitrary ECDSAKey where
+  arbitrary =
+    oneof
+      [ ECDSAKey ES256 <$> randomKey3,
+        ECDSAKey ES384 <$> randomKey3,
+        ECDSAKey ES512 <$> randomKey3
+      ]
+
+data PublicKey
+  = EdDSAPublicKey EdDSAKey
+  | ECDSAPublicKey ECDSAKey
+  deriving (Show, Eq)
+
+instance Arbitrary PublicKey where
+  arbitrary = oneof [EdDSAPublicKey <$> arbitrary, ECDSAPublicKey <$> arbitrary]
+
+-- | The credential public key encoded in COSE_Key format, as defined in Section 7
+-- of [RFC8152], using the CTAP2 canonical CBOR encoding form. The
+-- COSE_Key-encoded credential public key MUST contain the "alg" parameter and
+-- MUST NOT contain any other OPTIONAL parameters. The "alg" parameter MUST
+-- contain a COSEAlgorithmIdentifier value. The encoded credential public key
+-- MUST also contain any additional REQUIRED parameters stipulated by the
+-- relevant key type specification, i.e., REQUIRED for the key type "kty" and
+-- algorithm "alg" (see Section 8 of [RFC8152]).
+--
+-- Furthermore: CBOR values are CTAP2 canonical encoded.
+-- https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-client-to-authenticator-protocol-v2.0-id-20180227.html#ctap2-canonical-cbor-encoding-form
+decodePublicKey :: Decoder s PublicKey
+decodePublicKey = do
+  _n <- CBOR.decodeMapLenCanonical
+  ktyKey <- CBOR.decodeIntCanonical
+  when (ktyKey /= 1) $ fail "Expected required `kty`"
+  kty <- CBOR.decodeIntCanonical
+  case kty of
+    1 -> EdDSAPublicKey <$> decodeEdDSAKey
+    2 -> ECDSAPublicKey <$> decodeECDSAPublicKey
+    x -> fail $ "unexpected kty: " ++ show x
+
+decodeEdDSAKey :: Decoder s EdDSAKey
```

Comment:
All of the other decoders have `PublicKey` in the name, but this one just `Key`. But “key” here does refer to the public key, right? To avoid confusion, I would call this one `decodeEdDSAPublicKey`.